stdlib/public/core/KeyPath.swift

```diff
@@ -35,6 +35,8 @@ internal func _abstract(
 /// type.
 @_objcRuntimeName(_TtCs11_AnyKeyPath)
 public class AnyKeyPath: Hashable, _AppendKeyPath {
+  internal var _isPureStructKeyPath: Bool?
+  internal var _pureStructValueOffset: Int = 0
```

Comment:
Although the `AnyKeyPath` type layout is strictly speaking private, I am concerned that there are clients in practice who interpret the current object layout for reflection purposes, so changing the object layout by adding fields might be a binary compatibility issue. Memory usage is also a concern for key paths—there might be a lot of key path objects in an app, and adding a field to AnyKeyPath increases the memory usage of every one in the system.

If we compute the struct value offset during key path pattern instantiation, then the tribool shouldn't be necessary. It should also be true that no key path with a struct value offset ever has a KVC compatibility string (since only ObjC objects have those). So maybe we can overload the `_kvcCompatibilityString` pointer field to store the struct offset for pure struct key paths?

Reply:
Did you mean `_kvcKeyPathStringPtr`? I've grep'ed the entire compiler tree and I haven't found `_kvcKeyPathStringPtr` outside of KeyPath.swift. 
Is the proposal to store the value of the offset as if it were a pointer, like this?
`_kvcKeyPathStringPtr = UnsafePointer<CChar>(bitPattern: 0x08)`. This would represent an 8-byte offset. Any other proposal I can think of at the moment uses additional memory, including storing the pointer to a string (or Int) representing the actual offset, and using an enum/case where one case is an `UnsafePointer<CChar>` and the other is an Int offset. The latter happens to have a MemoryLayout size of 9 and offset of 16.

Reply:
One way of computing the offset without using extra memory is:
```
let offset = UnsafePointer<CChar>(bitPattern: 0x32) // _kvcKeyPathStringPtr
let offsetBase = UnsafePointer<CChar>(bitPattern: 0x01) //constant offset, can't be 0.
print(offsetBase!.distance(to: offset!) - 1) //48 (bytes)
```